### ISSUE TITLE
set_label() is ignored after subset()?

### ISSUE THREAD

**Issue by aldanor** [Microsoft/LightGBM#1197]
It seems like setting the label on a subset has no effect, here's a simple example:

```python
import numpy as np
import lightgbm as lgb

np.random.seed(0)
X = np.random.random((1000, 3))
y = np.random.random(1000) < 0.25

base = lgb.Dataset(X)                           # labels are not set here
train = base.subset(np.arange(500).construct()  # take the first 500 elements
train.set_label(y[:500])                        # set the label on the subset

lgb.train({}, train, 1)
```

It seems like although the label is set explicitly on `train`, it still tries to use the default (zero) labels from `base`:

```sh
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
```

**Comment by guolinke** [comment]
@aldanor 
you can try following code to get label from cpp side:
```
import numpy as np
import lightgbm as lgb

np.random.seed(0)
X = np.random.random((1000, 3))
y = np.random.random(1000) < 0.25

base = lgb.Dataset(X)                           # labels are not set here
train = base.subset(np.arange(500)).construct()  # take the first 500 elements
train.set_label(y[:500])                        # set the label on the subset
train.get_field('label')
```

And you can see the label is set correctly.

For the training stop, you may need to set `min_data=1`
```
lgb.train({'min_data':1}, train, 1)
```

**Comment by aldanor** [comment]
Thanks. Hmm.. so the warning because it uses the bin mapper from the base dataset and cannot find data for one of the 255 bins in the subset, or something like that? I've tried passing `'min_data': 1` but it doesn't seem to make it shut up.

**Comment by guolinke** [comment]
the reason of that warning:
1. small training data with large `min_data` (full name: `min_data_per_leaf`) and large `num_leaves` (`min_data * num_leaves > #data`). As a result, the tree cannot split any more.
2. the data is hard to split ( split cannot reduce loss). Since you using random data, so it is possible caused by this reason.
